### ISSUE TITLE
set! creates new bindings

### ISSUE THREAD

**Issue by ak-1** [cisco/ChezScheme#850]
My expectation was that
```
set! does not establish a new binding for var but rather alters the value of an existing binding.
```
However:
```
$ scheme
Chez Scheme Version 10.0.0
Copyright 1984-2024 Cisco Systems, Inc.

> x
Exception: variable x is not bound
Type (debug) to enter the debugger.
> (set! x 123)
> x
123
> (let () y)
Exception: variable y is not bound
Type (debug) to enter the debugger.
> (let () (set! y 456) y)
456
> 
```

**Comment by jltaylor-us** [comment]
There are several differences between the [Interaction Environment](https://cisco.github.io/ChezScheme/csug10.0/csug.html) and RnRS environments.  This is one of them.

> All identifiers not bound in the initial interaction environment and not defined by the programmer are treated as "potentially bound" as variables to facilitate the definition of mutually recursive procedures.

**Comment by ak-1** [comment]
```
All identifiers not bound in the initial interaction environment and not defined by the programmer are treated as "potentially bound" as variables to facilitate the definition of mutually recursive procedures.
```
After reading this I still would not expect this `set!` behavior.

I looked at various other scheme implementations. At least racket, guile and mit-scheme all allow mutually recursive definitions like `yin` / `yang` in their interaction environments, but do not allow to `set!` an unbound variable.

If this is really intended in ChezScheme, maybe it should be stated explicitly in the user guide.

**Comment by LiberalArtist** [comment]
> At least racket, &hellip; [does] not allow to `set!` an unbound identifier.

This is true by default, but can be controlled using the [`compile-allow-set!-undefined`](https://docs.racket-lang.org/reference/eval.html#%28def._%28%28quote._~23~25kernel%29._compile-allow-set%21-undefined%29%29) parameter.

**Comment by soegaard** [comment]
Time to close?
